src/lyra/abstract_domains/string/character_domain.py

```diff
@@ -0,0 +1,263 @@
+        """
+        certainly = self.certainly.union(other.certainly)
+        maybe = self.certainly.union(other.certainly)
+        return self._replace(certainly, maybe)
+
+
+class CharacterState(Store, State):
+
+    """
+        An implementation of the Character Inclusion Domain.
+
+        Maps from every string variable to a tuple of two sets: one set of characters that
+        are **certainly** included in the string and a set of characters that are **maybe**
+        included in the string. The string cannot contain characters from outside the **maybe**
+        set and must contain all characters in **certainly**. Therefore,
+        the **certainly** set is a subset of the **maybe** set.
+
+    """
+
+    class ConditionEvaluator(ExpressionVisitor):
+
+        def visit_Literal(self, expr, state=None, eval=None):
+            if expr.typ == StringLyraType():
+                return CharacterLattice(set(expr.val), set(expr.val))
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_Input(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_VariableIdentifier(self, expr, state=None, eval=None):
+            return state.store[expr]
+
+        def visit_LengthIdentifier(self, expr):
+            pass
+
+        def visit_ListDisplay(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_Range(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_AttributeReference(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_Subscription(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_Slicing(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_UnaryArithmeticOperation(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_UnaryBooleanOperation(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_BinaryArithmeticOperation(self, expr, state=None, eval=None):
+            if expr.operator == BinaryArithmeticOperation.Operator.Add:
+                left = self.visit(expr.left, state, eval)
+                right = self.visit(expr.right, state, eval)
+                return left.concat(right)
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
```

Comment:
```
            if expr in evaluation:
                return evaluation  # nothing to be done
            evaluated1 = self.visit(expr.left, state, evaluation)
            evaluated2 = self.visit(expr.right, state, evaluated1)
            if expr.operator == BinaryArithmeticOperation.Operator.Add:
                evaluated2[expr] = deepcopy(evaluated2[expr.left]).concat(evaluated2[expr.right])
                return evaluated2
```

You can probably just default to top otherwise.